### ISSUE TITLE
Add I2C support for whole STM32 family

### ISSUE THREAD

**Issue by strongly-typed** [modm-io/modm#101]
There has been some work done in roboterclubaachen/xpcc#350 by @rleh 

The STM32L4 driver (which is an ugly name, blame @strongly-typed), seems to be suitable also for F0, F3 and F7. I discussed with @chris-durand on Friday. I have F3 hardware here to test this.

There is the distinction between
* stm32
* stm32-extended
* stm32l4
in  [modm-devices](https://github.com/modm-io/modm-devices/blob/develop/tools/generator/dfg/stm32/stm_peripherals.py#L298).

How integrate this to modm?

**Comment by chris-durand** [comment]
I looked at the reference manuals yesterday. `stm32-extended` and `stm32l4` are in fact the same implementation. The memory maps and features seem to be completely identical except for the `WUPEN` bit (Wakeup from Stop mode enable) in `I2C_CR1` that is missing on some controllers (some F0, F7 ?).

I would suggest to rename the L4 I2C to `stm32-extended` and enable it for all the other controllers. That would need some changes in the L4 I2C driver and device file generator. The name `stm32-extended` is problably ok because it supports more features than the other I2C IP (Fast Mode Plus support).

**Comment by chris-durand** [comment]
It's not that easy, thank you ST:
![image](https://user-images.githubusercontent.com/25187160/46915807-36c76b80-cfb1-11e8-8074-3ad127525e35.png)

I hope this is somewhere in the CubeMX data… 
However, we do not support any of these optional features right now.

**Comment by rleh** [comment]
![STM32F0 reference manual: I2C register CR1](https://user-images.githubusercontent.com/2820734/46915816-71c99f00-cfb1-11e8-9906-207aa3fe1bf6.png)
The `WUPEN` bit is not used by the modm anyway, so we can simply ignore it.

I remember that the digital noise filter `DNF` is not available on all controllers.

**Comment by chris-durand** [comment]
> I remember that the digital noise filter DNF is not available on all controllers.

I think this is only true for the other IP. If I didn't overlook anything the `DNF` bits are available in all controllers that have the extended IP.

**Comment by chris-durand** [comment]
@rleh Are you currently working on this? It wouldn't be very efficient if we end up with 2 PRs implementing the same thing.

**Comment by rleh** [comment]
![image](https://user-images.githubusercontent.com/2820734/46915882-9a05cd80-cfb2-11e8-96ab-c5e357ef2cb9.png)
![image](https://user-images.githubusercontent.com/2820734/46915883-a722bc80-cfb2-11e8-9e3b-12fa8a7654e8.png)
![image](https://user-images.githubusercontent.com/2820734/46915984-72b00000-cfb4-11e8-84aa-f6c6f4198ede.png)
![image](https://user-images.githubusercontent.com/2820734/46915986-83f90c80-cfb4-11e8-9363-17097252e7eb.png)
![image](https://user-images.githubusercontent.com/2820734/46916006-c0c50380-cfb4-11e8-9f37-73c17b0e4016.png)

At least the feature tables are similar in the reference manuals...

**Comment by rleh** [comment]
CubeMX has data about e.g. `FastModePlus`:
```xml
<RefParameter Name="GPIO_FM6" Comment="Fast Mode" Type="list" DefaultValue="__NULL">
	<PossibleValue Value="__NULL" Comment="Disable"/>
	<PossibleValue Value="I2C_FASTMODEPLUS_PB6" Comment="Enable"/>
</RefParameter>
```
```xml
<RefParameter Comment="I2C Speed Mode " DefaultValue="I2C_Standard" Group="Timing configuration" Name="I2C_Speed_Mode" Type="list">
	<PossibleValue Comment="Standard Mode" Value="I2C_Standard"/>
	<PossibleValue Comment="Fast Mode" Value="I2C_Fast"/>
	<PossibleValue Comment="Fast Mode Plus" Semaphore="$IpInstance_FastMP" Value="I2C_Fast_Plus"/> 
</RefParameter>
```
But I don't know where `$IpInstance_FastMP` is defined :unamused:

**Comment by rleh** [comment]
> Are you currently working on this? It wouldn't be very efficient if we end up with 2 PRs implementing the same thing.

No.

**Comment by chris-durand** [comment]
This data is available because for `FastModePlus` the io pin has to be configured in `SYSCFG_CFGR1` to a special mode that can drive the I2C bus with 20mA.

**Comment by chris-durand** [comment]
According to CubeMX every device with the `stm32-extended` IP supports FM+ except I2C instance 2 on devices that match `STM32F030C8Tx|STM32F030R8Tx|STM32F051R8Tx|STM32F051C8Tx|STM32F051R8Ux|STM32F051C8Ux|STM32F0x0_Value_Line`.

All instances of `stm32-extended` support SMBus except I2C2 on F0/L0.

This data is somehow contradictory to the reference manuals. STM32F05xxx have no FM+ but I can configure it in CubeMX for STM32F058x8. The correct rule according to the reference manual is that every instance supports FM+ except for I2C2 on F03x, F05x and F070xB devices.

**Comment by strongly-typed** [comment]
Is there a way in lbuild to see if there are multiple instances of the peripheral?

in [module.lb](https://github.com/modm-io/modm/blob/develop/src/modm/platform/i2c/stm32l4/module.lb#L46):
```
properties["aid"] = "2" if self.instance == 4 else "1"
```

but ST sometimes leaves out the postfix `1` if there is only one peripheral.

Or should I just create an alias between `APB1RSTR` and `APB1RSTR1`?

**Comment by salkinium** [comment]
> Is there a way in lbuild to see if there are multiple instances of the peripheral?

You mean like this?
```py
has_instances = "instance" in device.get_driver("i2c")
```

This is just a dictionary, so you can always use the `get(key, default)` function:  `device.get_driver("i2c").get("instance", [])`.

**Comment by salkinium** [comment]
The above is an issue in the CAN drivers, where there's CAN or  CAN1 and/or CAN2. We follow the reference manual naming here, since we don't want to rename stuff (that gets you into trouble).

**Comment by chris-durand** [comment]
> in [module.lb](https://github.com/modm-io/modm/blob/develop/src/modm/platform/i2c/stm32l4/module.lb#L46):
> 
> ```
> properties["aid"] = "2" if self.instance == 4 else "1"
> ```
> 
> but ST sometimes leaves out the postfix `1` if there is only one peripheral.
> 
> Or should I just create an alias between `APB1RSTR` and `APB1RSTR1`?

This is probably not the right solution to this problem. This relation only holds for some L4. For instance, STM32F745 has 4 I2Cs but no `APB1RSTR1`.

**Comment by chris-durand** [comment]
> Or should I just create an alias between APB1RSTR and APB1RSTR1?

L4 controllers have `APB1RSTR1`/`APB1RSTR2`, all the others `APB1RSTR`. That should be manageable.

**Comment by strongly-typed** [comment]
[There](https://github.com/modm-io/modm/blob/develop/src/modm/platform/i2c/stm32l4/i2c_master.hpp.in#L82) is a little FIXME we/I need to fix. Blame me :-O

The timing blob was taken from the CubeMX. Does anybody know how to calculate these settings? Is a LUT with a bunch of common frequencies an intermediate solution?

**Comment by salkinium** [comment]
Closed in #118.
